Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -19,10 +21,40 @@ namespace BuildXL.Cache.ContentStore.Service.Grpc
     /// An implementation of a CAS copy helper client based on GRPC.
     /// TODO: Consolidate with GrpcClient to deduplicate code. (bug 1365340)
     /// </summary>
-    public class GrpcCopyClient : IShutdown<BoolResult>
+    public sealed class GrpcCopyClient : IShutdown<BoolResult>
     {
+        private static ConcurrentDictionary<(string, int), GrpcCopyClient> _clientPool = new ConcurrentDictionary<(string, int), GrpcCopyClient>();

```

Comment:
>string [](start = 45, length = 6)

Note that creating a custom type that defines its own GetHashCode and Equals is usually faster than defaults. (In fact, it may default to reference equality, which will just mean you will never find a dictionary entry. Check this.)

Reply:
System.ValueTuple is a struct with correct value semantic, so this should be totally fine.

---
In reply to: [271089260](https://github.com/Microsoft/BuildXL/pull/66#discussion_r271089260) [](ancestors = 271089260)